concurrent_test.go

```diff
@@ -0,0 +1,325 @@
+		testDuration     time.Duration
+	}{
+		{
+			name:             "1 reader",
+			readerCount:      1,
+			minReadInterval:  50 * time.Millisecond,
+			maxReadInterval:  100 * time.Millisecond,
+			minWriteInterval: 10 * time.Millisecond,
+			maxWriteInterval: 20 * time.Millisecond,
+			minWriteBytes:    200,
+			maxWriteBytes:    8000,
+			testDuration:     30 * time.Second,
+		},
+		{
+			name:             "10 readers",
+			readerCount:      10,
+			minReadInterval:  50 * time.Millisecond,
+			maxReadInterval:  100 * time.Millisecond,
+			minWriteInterval: 10 * time.Millisecond,
+			maxWriteInterval: 20 * time.Millisecond,
+			minWriteBytes:    200,
+			maxWriteBytes:    8000,
+			testDuration:     30 * time.Second,
+		},
+		{
+			name:             "50 readers",
+			readerCount:      50,
+			minReadInterval:  50 * time.Millisecond,
+			maxReadInterval:  100 * time.Millisecond,
+			minWriteInterval: 10 * time.Millisecond,
+			maxWriteInterval: 20 * time.Millisecond,
+			minWriteBytes:    500,
+			maxWriteBytes:    8000,
+			testDuration:     30 * time.Second,
+		},
+		{
+			name:             "100 readers",
+			readerCount:      100,
+			minReadInterval:  50 * time.Millisecond,
+			maxReadInterval:  100 * time.Millisecond,
+			minWriteInterval: 10 * time.Millisecond,
+			maxWriteInterval: 20 * time.Millisecond,
+			minWriteBytes:    500,
+			maxWriteBytes:    8000,
+			testDuration:     30 * time.Second,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			concurrentReadAndWrite(t,
+				bucket,
+				keys,
+				tc.readerCount,
+				tc.minReadInterval, tc.maxReadInterval,
+				tc.minWriteInterval, tc.maxWriteInterval,
+				tc.minWriteBytes, tc.maxWriteBytes,
+				tc.testDuration)
+		})
+	}
+}
+
+func concurrentReadAndWrite(t *testing.T,
+	bucket []byte,
+	keys []string,
+	readerCount int,
+	minReadInterval, maxReadInterval time.Duration,
+	minWriteInterval, maxWriteInterval time.Duration,
+	minWriteBytes, maxWriteBytes int,
+	testDuration time.Duration) {
+
+	// prepare the db
+	db := btesting.MustCreateDB(t)
+	err := db.Update(func(tx *bolt.Tx) error {
+		_, err := tx.CreateBucket(bucket)
+		return err
+	})
+	require.NoError(t, err)
+
+	stopCh := make(chan struct{}, 1)
+	errCh := make(chan error, readerCount+1)
+
+	// start readonly transactions
+	g := new(errgroup.Group)
+	for i := 0; i < readerCount; i++ {
+		reader := &readWorker{
+			db:              db,
+			bucket:          bucket,
+			keys:            keys,
+			minReadInterval: minReadInterval,
+			maxReadInterval: maxReadInterval,
+			errCh:           errCh,
+			stopCh:          stopCh,
+			t:               t,
+		}
+		g.Go(reader.run)
+	}
+
+	// start write transaction
+	writer := writeWorker{
+		db:               db,
+		bucket:           bucket,
+		keys:             keys,
+		minWriteBytes:    minWriteBytes,
+		maxWriteBytes:    maxWriteBytes,
+		minWriteInterval: minWriteInterval,
+		maxWriteInterval: maxWriteInterval,
+
+		errCh:  errCh,
```

Comment:
not blocking: 
  I would probably pass the operational data to writer.run() directly. 

```
g.Go(func() { return writer.run(ctx, t)}) 
```
or
```
g.Go(writer.runCallback(ctx, t))
```
  - ctx replaces stopCh. 
  - errCh not needed as function called by `Go` can return error and automatically cancel context.

Reply:
> errCh not needed as function called by Go can return error and automatically cancel context.

We need to stop the test when any worker (including reader and writer) runs into error, see line 153 ~ 156.

```
	select {
	case <-time.After(testDuration):
	case <-errCh:
	}
```

Others comments seem not a big deal.